soroban-sdk/src/testutils.rs

```diff
@@ -29,6 +31,70 @@ pub trait Ledger {
         F: FnMut(&mut LedgerInfo);
 }
 
+pub mod budget {
+    use core::fmt::Display;
+
+    #[doc(inline)]
+    pub use crate::env::internal::budget::CostType;
+
+    /// Budget that tracks the resources consumed for the environment.
+    ///
+    /// ### Examples
+    ///
+    /// ```
+    /// use soroban_sdk::{Env, Symbol};
+    ///
+    /// # #[cfg(feature = "testutils")]
+    /// # fn main() {
+    /// #     let env = Env::default();
+    /// env.budget().reset();
+    /// // ...
+    /// println!("{}", env.budget());
+    /// # }
+    /// # #[cfg(not(feature = "testutils"))]
+    /// # fn main() { }
+    /// ```
+    pub struct Budget(crate::env::internal::budget::Budget);
+
+    impl Display for Budget {
+        fn fmt(&self, f: &mut core::fmt::Formatter<'_>) -> core::fmt::Result {
+            writeln!(f, "CPU Instructions: {}", self.get_cpu_instruction_count())?;
+            writeln!(f, "Memory Bytes: {}", self.get_memory_bytes_count())?;
```

Comment:
Any work that is done by via the host (calling host function, accessing storage etc) are accurate. The main difference when running non-WASM (native) contract is the `WasmInsnExec`, `WasmMemAlloc`  and `Vm*` ones will not be counted, because there is no Vm running. Therefore the cost numbers will underestimate the actual contract running cost. But given most of the heavy lifting is done by the host anyway, they should be close enough to give the user some sense of the actual cost.

Reply:
I'll add notes about this in the docs.

Reply:
I added some docs explaining this in 51cb8fe3.

Reply:
would it be possible to detect when it is running non-wasm contract and print a notice to the user?

Reply:
I don't think so. Also, a test might interact with both wasm and rust invocations. They can be mixed.